Microsoft.Azure.CosmosRepository/src/Providers/DefaultCosmosUniqueKeyPolicyProvider.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) IEvangelist. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Collections.Concurrent;
+using Microsoft.Azure.CosmosRepository.Attributes;
+using Microsoft.Azure.Cosmos;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.Azure.CosmosRepository.Providers
+{
+    class DefaultCosmosUniqueKeyPolicyProvider : ICosmosUniqueKeyPolicyProvider
+    {
+
+        static readonly Type _uniqueKeyAttributeType = typeof(UniqueKeyAttribute);
+        static readonly ConcurrentDictionary<Type, UniqueKeyPolicy> _uniqueKeyPolicyMap =
+            new ConcurrentDictionary<Type, UniqueKeyPolicy>();
+
+        /// <inheritdoc />
+        public UniqueKeyPolicy GetUniqueKeyPolicy<TItem>() where TItem : IItem =>
+            _uniqueKeyPolicyMap.GetOrAdd(typeof(TItem), GetPartitionKeyNameFactory);
+
+        static UniqueKeyPolicy GetPartitionKeyNameFactory(Type type)
+        {
+            Dictionary<string, List<PropertyInfo>> dict = new Dictionary<string, List<PropertyInfo>>();
+            IEnumerable<PropertyInfo> uniqueKeyed = type.GetProperties().Where((p) => Attribute.IsDefined(p, _uniqueKeyAttributeType));
+
+            foreach (PropertyInfo property in uniqueKeyed)
+            {
+                string keyName = property.GetCustomAttribute<UniqueKeyAttribute>().KeyName;
+                if (!dict.ContainsKey(keyName))
+                {
+                    dict.Add(keyName, new List<PropertyInfo>());
+                }
+                dict[keyName].Add(property);
+            }
+
+            UniqueKeyPolicy policy = new UniqueKeyPolicy();
+            foreach (KeyValuePair<string, List<PropertyInfo>> kv in dict)
+            {
+                UniqueKey uniqueKey = new UniqueKey();
+                foreach (PropertyInfo property in kv.Value)
+                {
+                    uniqueKey.Paths.Add("/" + property.Name);
+                }
+                policy.UniqueKeys.Add(uniqueKey);
+            }
+
+            return policy;
+        }
+

```

Comment:
```suggestion
            Dictionary<string, List<PropertyInfo>> dict = new();
            IEnumerable<PropertyInfo> uniqueKeyed = type.GetProperties().Where((p) => Attribute.IsDefined(p, _uniqueKeyAttributeType));

            foreach (PropertyInfo property in uniqueKeyed)
            {
                string keyName = property.GetCustomAttribute<UniqueKeyAttribute>().KeyName;
                if (!dict.ContainsKey(keyName))
                {
                    dict.Add(keyName, new List<PropertyInfo>());
                }
                dict[keyName].Add(property);
            }

            UniqueKeyPolicy policy = new();
            foreach (KeyValuePair<string, List<PropertyInfo>> kv in dict)
            {
                UniqueKey uniqueKey = new();
                foreach (PropertyInfo property in kv.Value)
                {
                    uniqueKey.Paths.Add($"/{property.Name}");
                }
                policy.UniqueKeys.Add(uniqueKey);
            }

            return policy;
        }
```

I think before we ship this - I'll want to run a few more tests on it. It seems like this is something that could be optimized a bit more as well. Also, shouldn't be possible for there to be no policy when there are no attributes used on an item?

Reply:
That is true. As is now I think it would implicitly do so by providing a policy with no keys. But maybe it would be better to return either null or maybe check if there are any unique keys to beginn with.